mesonbuild/interpreter/interpreter.py

```diff
@@ -1135,6 +1135,9 @@ def func_project(self, node: mparser.FunctionNode, args: T.Tuple[str, T.List[str
         if self.build.project_version is None:
             self.build.project_version = self.project_version
         proj_license = kwargs['license']
+        if len(proj_license) > 1:
+            FeatureDeprecated.single_use('license as an array of strings', '0.62.0', self.subproject,
```

Comment:
I wonder if we actually need a version-gated feature? We recommend against it purely because it is "a bad idea" and the alternative is to do something that has also worked since this kwarg was added. So there's no reason to only recommend against it once you bump your minimum meson.build high enough.

Reply:
You're right, the way this is written is bad. I've changed it so that the function takes a `str | list[str]`, and that passing `list[str]` is deprecated, and the warning is gated on that.

Reply:
While this does fine tune the check to also trigger on `license: ['GPL-2.0-only']` it doesn't really answer my question, which is whether we should use `mlog.deprecation` and print warnings even for projects that support Meson 0.45, instead of using FeatureDeprecated and only warning people away from this once they bump `meson_version: '>=0.62'`.

Again, the point of a FeatureDeprecated is typically because we change how things work and add new features which replace old and badly designed features, so as a result a FeatureDeprecated does not warn you if you cannot migrate to the new method due to your requirement to support old versions of Meson.

Reply:
I've made a change to add a warning in the even that you're using a version < 0.62 in addition to the deprecation warning (either the FeatureDeprecated or the separate warning). The FeatureDeprecated is still useful as passing an array should go away, but you're right, encouraging the use of an SPDX license expression even in older versions is desirable

Reply:
This seems like a very complicated way to add two distinct warnings for one thing, when passing as an array is *still* not something that needs to version-gated the warning.

Deprecating a feature with a replacement that has been available for as long as the feature itself, doesn't need a version gate, whether that feature is "multiple license strings or "passing as an array".

We can just tell people using Meson 0.45 "you know, it's kind of bad to not use SPDX, and since you're going to use SPDX, it's kind of silly to pass a single string wrapped in an array, and also this still works on 0.45 *anyway* (and quite frankly it works on like 0.25 also)".

Reply:
I'm strongly against removing the `FeatureDeprecated`, because it provides a clean way for projects receiving bug reports about the deprecation to know exactly what's happening, and when it was deprecated. Having been on the end of CMake policy bug reports/PRs that don't include that information, I think this makes for good user experience. It's immaterial that the replacement has been available since the beginning, 0.62 is when we declared "this is going away".

The other problem we have, is that we don't (and can't easily without external deps) verify that the project license is an SPDX expression, only that it's an array or a single string. That string could easily be `LOLZ!1!`, and we can't say with certainty "this is not SPDX!". But we can say that `['GPL', 'MIT']` is ambiguous, knock that off.

Reply:
That would imply we should extend FeatureCheck with a `.single_use(..., always_warn=True)` option, and just emit the FeatureDeprecated but ensure it is always logged as a full warning.

Reply:
@dcbaker could you update your PR to add FeatureDeprecated.single_use(..., always_warn=True) please?